crates/rpc/src/jsonrpc/websocket/logic.rs

```diff
@@ -437,6 +471,89 @@ async fn event_subscription(
+                    (FinalityStatus::NotReceived, _) => {
+                        // "NOT_RECEIVED" status is never sent to the client.
+                        if start.elapsed() > timeout {
+                            // The transaction was not found on the gateway after some time.
+                            // This means the transaction is probably not valid.
+                            msg_sender
+                                .send(ResponseEvent::RpcError(RpcError::ApplicationError(
+                                    ApplicationError::SubscriptionTransactionHashNotFound {
+                                        transaction_hash,
+                                        subscription_id,
+                                    },
+                                )))
+                                .await
+                                .ok();
+                            break;
+                        }
+                        None
+                    }
+                    (FinalityStatus::Received, _) => Some(TransactionStatusUpdate::Received),
+                    (
+                        FinalityStatus::AcceptedOnL1 | FinalityStatus::AcceptedOnL2,
+                        ExecutionStatus::Succeeded,
+                    ) => Some(TransactionStatusUpdate::Succeeded),
+                    (
+                        FinalityStatus::AcceptedOnL1 | FinalityStatus::AcceptedOnL2,
+                        ExecutionStatus::Reverted,
+                    ) => Some(TransactionStatusUpdate::Reverted),
+                };
+                let status_changed = match (last_status, update) {
+                    (Some(last), Some(update)) => last < update,
+                    (None, Some(_)) => true,
+                    (Some(_), None) => false,
+                    (None, None) => false,
+                };
+                if status_changed {
+                    // Status changed, send an update to the client.
+                    last_status = update;
+                    if msg_sender
+                        .send(ResponseEvent::TransactionStatus(SubscriptionItem {
+                            subscription_id,
+                            item: Arc::new(update.unwrap()),
+                        }))
+                        .await
+                        .is_err()
+                    {
+                        break;
+                    }
+                }
+                if tx_status.execution_status == ExecutionStatus::Rejected
+                    || tx_status.finality_status == FinalityStatus::AcceptedOnL1
+                    || tx_status.finality_status == FinalityStatus::AcceptedOnL2
+                {
+                    // Final status reached, close the subscription.
+                    break;
+                }
+            }
+            Err(e) => {
+                tracing::warn!(%transaction_hash, %e, "Failed to poll transaction status");
```

Comment:
There is a weird edge case where if the gateway is down or unavailable for a lenghty period of time.. what should we do in those cases? Currently we just keep polling, which is maybe adequate..? The only alternative I can think of is also tracking how many consecutive failures we get, and then bail after say 5 failed polls with "Feeder gateway unavailable" or something.

Does not have to be included in this PR if you don't think its worth adding - I'm unsure about the pros/cons myself.

My concern is:

1. User subscribes
2. Gateway is down / unavailable
3. User gets zero feedback for a long time

Ideally they should get an error indicating a connection issue.

A similar issue presents itself if the gateway is down, but we already know about the transaction locally. This is a flaw with our simple approach of just polling the gateway instead of relying on local data as well.

I suggest

1. Add this consecutive failure counter with a custom error output to this or a follow-up PR.
2. A follow-up issue to more carefully design a local + gateway poll combination which is both simple and robust.